File: typescript/packages/bench/src/reports/orderThroughputReport.ts
/** Completed against still-open order counts. */

import type { ReportRow } from "../contracts/reportContract";
import type { Dataset } from "../dataset";
import { reportRow } from "../contracts/reportContract";
import { AbstractReport } from "../support/abstractReport";

/** Completed against still-open order counts. */
export class OrderThroughputReport extends AbstractReport {
    /** Registry slug; also the URL segment. */
    static readonly SLUG = "order-throughput";

    constructor() {
        super(OrderThroughputReport.SLUG, "Order throughput");
    }

    override rows(data: Dataset): ReportRow[] {
        return [
            reportRow("completed", data.completedOrders().length),
            reportRow("open", data.openOrders().length),
        ];
    }
}

File: rust/crates/atelier-bench/src/reports/order_throughput_report.rs
//! Completed against still-open order counts.

use crate::contracts::report_contract::{ReportContract, ReportRow};
use crate::dataset::Dataset;
use crate::support::abstract_report::ReportBase;

/// Completed against still-open order counts.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct OrderThroughputReport {
    base: ReportBase,
}

impl OrderThroughputReport {
    /// Registry slug; also the URL segment.
    pub const SLUG: &'static str = "order-throughput";

    pub const fn new() -> Self {
        Self {
            base: ReportBase::new(Self::SLUG, "Order throughput"),
        }
    }

    /// Shared identity and rendering precision.
    pub const fn base(&self) -> &ReportBase {
        &self.base
    }
}

impl Default for OrderThroughputReport {
    fn default() -> Self {
        Self::new()
    }
}

impl ReportContract for OrderThroughputReport {
    fn slug(&self) -> &'static str {
        self.base.slug()
    }

    fn title(&self) -> &'static str {
        self.base.title()
    }

    fn rows(&self, data: &Dataset) -> Vec<ReportRow> {
        vec![
            ReportRow::new("completed", data.completed_orders().len() as f64),
            ReportRow::new("open", data.open_orders().len() as f64),
        ]
    }
}
